File: framework/languages/dotnet/samples/Bingo/Server/Session/Sessions/BingoSession.cs
using Zlink.Framework.Contracts.Messaging;
using Zlink.Framework.Contracts.Streams;

namespace Bingo.Server.Session.Sessions;

internal sealed class BingoSession(
    IZLinkSessionContext context) : IZLinkSession
{
    public IZLinkSessionContext Context { get; } = context;

    public ValueTask OnConnectedAsync(CancellationToken cancellationToken)
    {
        return ValueTask.CompletedTask;
    }

    public async ValueTask OnDisconnectedAsync(CancellationToken cancellationToken)
    {
        foreach (var actor in Context.Actors.Bound) await actor.NotifyDisconnectedAsync(cancellationToken);
    }

    public ValueTask OnErrorAsync(
        ZLinkStreamError error,
        CancellationToken cancellationToken)
    {
        return ValueTask.CompletedTask;
    }

    public async ValueTask OnDispatchAsync(
        ZLinkSessionDispatchContext dispatch,
        ZLinkMessage payload,
        CancellationToken cancellationToken)
    {
        if (await Context.Handlers.TryHandleAsync(dispatch, payload, cancellationToken))
            return;

        var actor = RequireSingleBoundActor($"relaying packet '{dispatch.PacketName}'");
        await actor.RelayAsync(
                payload,
                cancellationToken)
            ;
    }

    private IZLinkSessionActor RequireSingleBoundActor(string action)
    {
        var actors = Context.Actors.Bound;
        return actors.Count switch
        {
            1 => actors.Single(),
            0 => throw new InvalidOperationException($"Client must authenticate before {action}."),
            _ => throw new InvalidOperationException($"Exactly one actor must be bound before {action}.")
        };
    }
}

File: framework/languages/node/samples/Bingo.Ts/Server/Session/Sessions/bingo-session.ts
import type {
  ZLinkMessage,
  ZLinkSession,
  ZLinkSessionContext,
  ZLinkSessionDispatchContext,
  ZLinkSessionFactory
} from '@zlink-systems/framework';

class BingoSession implements ZLinkSession {
  constructor(
    readonly context: ZLinkSessionContext
  ) {}

  async onDispatch(dispatch: ZLinkSessionDispatchContext, payload: ZLinkMessage, signal?: AbortSignal): Promise<void> {
    if (await this.context.handlers.tryHandle(dispatch, payload)) {
      return;
    }
    const actor = this.context.actors.bound.length === 1 ? this.context.actors.bound[0] : undefined;
    if (actor === undefined) {
      throw new Error(`Client must authenticate before relaying packet '${dispatch.packetName}'.`);
    }
    await actor.relay(payload, signal);
  }

  async onDisconnected(): Promise<void> {
    const actors = this.context.actors.bound;
    await Promise.allSettled(actors.map((actor) => actor.notifyDisconnected()));
    console.error(`bingo-lifecycle session-disconnect actor=${actors[0]?.actorId ?? '-'} destroy=false`);
  }
}

class BingoSessionFactory implements ZLinkSessionFactory<BingoSession> {
  async create(context: ZLinkSessionContext): Promise<BingoSession> {
    return new BingoSession(context);
  }
}

export { BingoSession, BingoSessionFactory };
